reference-implementation/lib/readable-byte-stream.js

```diff
@@ -0,0 +1,965 @@
+
+export default class ReadableByteStream {
+  constructor(underlyingByteSource = {}) {
+    this._state = 'readable';
+    this._reader = undefined;
+    this._storedError = undefined;
+
+    this._controller = undefined;
+    this._controller = new ReadableByteStreamController(this, underlyingByteSource);
+  }
+
+  cacnel(reason) {
+    if (IsReadableByteStream(this) === false) {
+      return Promise.reject(
+          new TypeError('ReadableByteStream.prototype.cancel can only be used on a ReadableByteStream'));
+    }
+
+    if (IsReadableByteStreamLocked(this)) {
+      return Promise.reject(new TypeError('Cannot cancel a stream that already has a reader'));
+    }
+
+    return CancelReadableByteStream(this, reason);
+  }
+
+  getByobReader() {
+    if (IsReadableByteStream(this) === false) {
+      throw new TypeError('ReadableByteStream.prototype.getByobReader can only be used on a ReadableByteStream');
+    }
+
+    return new ReadableByteStreamByobReader(this);
+  }
+
+  getReader() {
+    if (IsReadableByteStream(this) === false) {
+      throw new TypeError('ReadableByteStream.prototype.getReader can only be used on a ReadableByteStream');
+    }
+
+    return new ReadableByteStreamReader(this);
+  }
+}
+
+class ReadableByteStreamController {
+  constructor(controlledReadableByteStream, underlyingByteSource) {
+    if (IsReadableByteStream(controlledReadableByteStream) === false) {
+      throw new TypeError('ReadableByteStreamController can only be constructed with a ReadableByteStream instance');
+    }
+
+    if (controlledReadableByteStream._controller !== undefined) {
+      throw new TypeError(
+          'ReadableByteStreamController instances can only be created by the ReadableByteStream constructor');
+    }
+
+    this._controlledReadableByteStream = controlledReadableByteStream;
+
+    this._underlyingByteSource = underlyingByteSource;
+
+    this._considerReissueUnderlyingByteSourcePullOrPullInto = false;
+    this._insideUnderlyingByteSource = false;
+
+    this._pendingPullIntos = [];
+
+    this._queue = [];
+    this._totalQueuedBytes = 0;
+
+    InvokeOrNoop(underlyingByteSource, 'start', [this]);
+  }
+
+  close() {
+    if (!IsReadableByteStreamController(this)) {
+      throw new TypeError(
+          'ReadableByteStreamController.prototype.close can only be used on a ReadableByteStreamController');
+    }
+
+    const stream = this._controlledReadableByteStream;
+
+    if (stream._closeRequested) {
+      throw new TypeError('The stream has already been closed; do not close it again!');
+    }
+    if (stream._state !== 'readable') {
+      throw new TypeError('The stream is not in the readable state and cannot be closed');
+    }
+
+    if (this._totalQueuedBytes > 0) {
+      this._closeRequested = true;
+
+      return;
+    }
+
+    const reader = stream._reader;
+
+    if (reader === undefined || IsReadableByteStreamReader(reader)) {
```

Comment:
I think the naming of IsReadableByteStreamReader is a bit confusing, even though it mirrors the getReader/getByobReader methods. Maybe IsReadableByteStreamDefaultReader?